design/envoy-json-logging-custom-fields-design.md

```diff
@@ -0,0 +1,146 @@
+This mirrors current default translation table, but allows users to define custom translations.
+Validation could be done in two stages:
+  1. Validate that all the right-hand-sides of translations appear valid.
+  2. Validate that all entries in `json-fields` are in the default or user-defined translation tables.
+However, [#3033](https://github.com/projectcontour/contour/pull/3033) does not include validation logic.
+Having a user-defined translation table makes adding a field like `content-id` require changes in two places (one to define `content-id` in the translation table and one to add it to the `json-fields` list).
+
+In this alternative, the example configuration would be written as the following:
+```yaml
+json-fields:
+   - @timestamp
+   - method
+   - content-id
+   - response_duration
+extra-json-fields:
+   content-id: %REQ(X-CONTENT-ID)%
+   response_duration: %RESPONSE_DURATION%
+```
+
+### Map Syntax
+The proposed implementation would require parsing each value to split at the first `=`.
+Instead of using an `=` within the string to separate a key and value, the Contour configuration could be a YAML key and value.
+In this approach, the example configuration would look like the following:
+```yaml
+json-fields:
+   @timestamp:
+   method:
+   content-id: %REQ(X-CONTENT-ID)%
+   response_duration: %RESPONSE_DURATION%
+```
+
+This approach is not backward-compatible because the structure of `json-fields` is changing from an array to a map.
+Existing configuration would not work.
+The other drawback of this approach is aesthetic: it could be argued that empty values for the keys with default translations is ugly.
+
+## Security Considerations
+Misconfiguration of the property can leak unwanted information to the access log file.
+
+## Compatibility
+Both original and the alternative translation table high-level designs don't introduce breaking changes, but the map syntax alternative would be a breaking change.
+The current implementation skips unknown fields so any new custom fields will be ignored.
+
+## Open Issues
+
+* Should all the fields with obvious translations consistently either require parameterization or not?
+  - In the current implementation, `method` becomes `"method": "%REQ(:METHOD)%"`.
+  - When the new implementation adds a way to achieve the JSON config `"response_duration": "%RESPONSE_DURATION%"`, should the user write:
+    * `response_duration`, or
+    * `response_durationn=%RESPONSE_DURATION%`
+  - This is related to [#2523](https://github.com/projectcontour/contour/issues/2523)
+* Is one of the alternative approaches preferable?
+   - [User-defined Translation Table](#user-defined-translation-table) seems easier to implement because it was partially implemented in #3033 and because the configuration is more structured and easier to work with as there no need for additional string parsing.
+   - [Map Syntax](#map-syntax) is not backward-compatible, but does avoid string parsing.
```

Comment:
Map syntax seems clean, but we would need to introduce it in a new config field, then deprecate the old field after a number of releases. Would still need to define what happens to the built-in field mappings.

Reply:
Yea, since the map syntax would require deprecating the prior configuration field anyway, the new field would not have to support the built-in field mappings. In our example in this doc, we assumed that empty values would imply that the key should be looked up in the built-in field mappings, but I'm not sure whether that's desirable or not.

## In favor of a built-in translation table:
* Users of Contour only have to read Contour's docs (vs. also referencing Envoy's config syntax).
* Users get reasonable, standardized field names because that's what's setup by the field mappings.

## In favor of writing all key-value pairs out:
* Users of Contour can read the Envoy docs to decide what config to write. They don't need to know Contour's specific translation of special keys.

Reply:
I like the ease of use of having the translation table, personally. I think that having the two types seems like an okay tradeoff for that ease of use (for now at least).